### PR TITLE
Add task to download a static compiled socat binary

### LINKED ISSUE #1861: CoreOS is #NoSocat
So for some reason coreos hates socat :-)

@philk pointed this out in https://github.com/kubernetes/kops/pull/1813#issuecomment-278822176 , referencing https://github.com/coreos/bugs/issues/1114  It looks like one workaround would be to install socat.

@aledbf suggested we could also use hyperkube

If socat isn't easily available for coreos, we could also write our own just-enough-socat implementation in go, and avoid the binary problems ( `gocat` ).  But I need to check how we are using socat...

### PR DESCRIPTION
**Pull request by aledbf**
When installing CoreOS

fixes #1861

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kops/1895)

### CODE REVIEW

**Comment by aledbf** [comment]
@philk FYI

File: nodeup/pkg/model/kubelet.go [hunk 1]
```diff
@@ -35,6 +38,8 @@ type KubeletBuilder struct {
 
 var _ fi.ModelBuilder = &DockerBuilder{}
 
+const socatStaticBinaryURL = "https://github.com/aledbf/socat-static-binary/releases/download/v0.0.1/socat-linux-amd64"
```
Comment (chrislovecnm) [inline nodeup/pkg/model/kubelet.go]: We need a checksum as well to check file integrity :)

Comment (aledbf) [inline nodeup/pkg/model/kubelet.go]: Good point :)

Comment (aledbf) [inline nodeup/pkg/model/kubelet.go]: done

Comment (philk) [inline nodeup/pkg/model/kubelet.go]: Might be worth making the URL configurable like [NODEUP_URL](https://github.com/kubernetes/kops/blob/master/upup/pkg/fi/cloudup/urls.go#L61) but personally I'm fine with it this way.

Comment (aledbf) [inline nodeup/pkg/model/kubelet.go]: @philk good idea. Done

Comment (chrislovecnm) [inline nodeup/pkg/model/kubelet.go]: Not sure that travels through to `nodeup` ... never tried it

File: nodeup/pkg/model/kubelet.go [hunk 2]
```diff
@@ -105,6 +113,37 @@ func (b *KubeletBuilder) Build(c *fi.ModelBuilderContext) error {
...
+		}
+
+		socatURL := os.Getenv("SOCAT_URL")
+		if socatURL == "" {
+			socatURL = socatStaticBinaryURL
+		}
+		_, err = fi.DownloadURL(socatURL, "/opt/bin/socat", hash)
+		if err != nil {
+			return fmt.Errorf("unexpected error downloading socat from %v: %v", socatURL, err)
+		}
```
Comment (philk) [inline nodeup/pkg/model/kubelet.go]: Need executable:

```golang

		err = os.Chmod("/opt/bin/socat", 0755)
		if err != nil {
			return fmt.Errorf("failed setting executable on socat binary: %s", err)
		}
```

File: nodeup/pkg/model/kubelet.go [hunk 3]
```diff
@@ -105,6 +113,37 @@ func (b *KubeletBuilder) Build(c *fi.ModelBuilderContext) error {
 
 	c.AddTask(b.buildSystemdService())
 
+	if b.Distribution == distros.DistributionCoreOS {
+		sha1URL := os.Getenv("SOCAT_SHA1_URL")
```
Comment (philk) [inline nodeup/pkg/model/kubelet.go]: It's a SHA256, should probably change that (realizing that means you'll have to change it in the GH release too)

**Comment by philk** [comment]
Another issue, `/opt/bin/` isn't in the kubelet's `PATH`

```
$ sudo cat /proc/1532/environ
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/binJOURNAL_STREAM=8:19581DAEMON_ARGS=--allow-privileged=true --api-servers=http://127.0.0.1:8080 --babysit-daemons=true --cgroup-root=docker --cloud-provider=aws --cluster-dns=100.64.0.10 --cluster-domain=cluster.local --enable-debugging-handlers=true --eviction-hard=memory.available<100Mi,nodefs.available<10%,nodefs.inodesFree<5%,imagefs.available<10%,imagefs.inodesFree<5% --hostname-override=ip-10-0-39-78.us-west-2.compute.internal --network-plugin=cni --node-labels=kubernetes.io/role=master --non-masquerade-cidr=100.64.0.0/10 --pod-manifest-path=/etc/kubernetes/manifests --register-schedulable=false —v=2
```

I don't immediately have a good idea how to solve that one.

**Comment by aledbf** [comment]
@philk updated and kubelet PATH fixed.

**Comment by aledbf** [comment]
```console
...
 1500 ?        Ssl    0:32 /opt/kubernetes/bin/kubelet --allow-privileged=true --api-servers=https://api.internal.uswest2-02.xxxxxx.cl --babysit-daemons=true --cgroup-root=docker --cloud-provider=aws --cluster-dns=100.64.0.10 --c
 1536 ?        S      0:00  \_ journalctl -k -f
core@ip-172-20-57-66 ~ $ sudo cat /proc/1500/environ
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/opt/kubernetes/binJOURNAL_STREAM=8:20629DAEMON_ARGS=--allow-privileged=true --api-servers=https://api.internal.uswest2-02.xxxxxx.cl --babysit-daemons=true --cgroup-root=docker --cloud-provider=aws --cluster-dns=100.64.0.10 --cluster-domain=cluster.local --enable-debugging-handlers=true --eviction-hard=memory.available<100Mi,nodefs.available<10%,nodefs.inodesFree<5%,imagefs.available<10%,imagefs.inodesFree<5% --hostname-override=ip-172-20-57-66.us-west-2.compute.internal --network-plugin-mtu=9001 --network-plugin=kubenet --non-masquerade-cidr=100.64.0.0/10 --pod-manifest-path=/etc/kubernetes/manifests --v=2
```

**Comment by aledbf** [comment]
superseded by #1909

**Pull request closed without merging**